### PR TITLE
update pkg about database

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
### What problem does this PR solve? 
update some function about database

this pr is splited from pr: https://github.com/pingcap/tidb-tools/pull/130
make it easy to review

### What is changed and how it works?
1. add some function to get tidb's  buckets info
2. add some function about index

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

**Comment by WangXiangUSTC** [comment]
@GregoryIan @csuzhangxc PTAL

File: pkg/dbutil/common.go [hunk 1]
```diff
@@ -208,6 +213,52 @@ func GetRandomValues(ctx context.Context, db *sql.DB, schemaName, table, column
 	return randomValue, nil
 }
 
+// GetMinMaxValue get min and max value.
```
Comment (IANTHEREAL) [inline pkg/dbutil/common.go]: GetMinMaxValue return min and max value of given column by specified `limitRange` condition

File: pkg/dbutil/common.go [hunk 2]
```diff
@@ -415,9 +598,21 @@ func IsTiDB(ctx context.Context, db *sql.DB) (bool, error) {
 
 // TableName returns `schema`.`table`
 func TableName(schema, table string) string {
-	return fmt.Sprintf("`%s`.`%s`", escapeName(schema), escapeName(table))
+	return fmt.Sprintf("%s.%s", escapeName(schema), escapeName(table))
 }
 
 func escapeName(name string) string {
-	return strings.Replace(name, "`", "``", -1)
+	return fmt.Sprintf("`%s`", name)
```
Comment (IANTHEREAL) [inline pkg/dbutil/common.go]: why change it?

File: pkg/dbutil/common.go [hunk 3]
```diff
@@ -208,6 +213,52 @@ func GetRandomValues(ctx context.Context, db *sql.DB, schemaName, table, column
 	return randomValue, nil
 }
 
+// GetMinMaxValue return min and max value of given column by specified limitRange condition.
+func GetMinMaxValue(ctx context.Context, db *sql.DB, schema, table, column string, limitRange string, collation string, args []interface{}) (string, string, error) {
```
Comment (csuzhangxc) [inline pkg/dbutil/common.go:217]: is `args` needed here?

Comment (WangXiangUSTC) [reply pkg/dbutil/common.go:217]: limitRange may looks like "a > ? AND a < ?", need use args when query

File: pkg/dbutil/common.go [hunk 4]
```diff
@@ -320,6 +371,138 @@ func GetCRC32Checksum(ctx context.Context, db *sql.DB, schemaName, tableName str
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for rows.Next() {
+		var dbName, tableName, partitionName, columnName, lowerBound, upperBound sql.NullString
+		var isIndex, bucketID, count, repeats sql.NullInt64
+
+		// add partiton_name in new version
+		switch len(cols) {
```
Comment (csuzhangxc) [inline pkg/dbutil/common.go:413]: can get these info from `mysql`.`stats_buckets`? if so, use `SELECT NEEDED COLUMNS FROM` may have better compatibility.

File: pkg/dbutil/common.go [hunk 5]
```diff
@@ -320,6 +371,138 @@ func GetCRC32Checksum(ctx context.Context, db *sql.DB, schemaName, tableName str
...
+	}
+
+	return buckets, errors.Trace(rows.Err())
+}
+
+// AnalyzeValuesFromBuckets analyze upperBound or lowerBound to string for each column.
+// upperBound and lowerBound are looks like '(123, abc)' for multiple fields, or '123' for one field.
+func AnalyzeValuesFromBuckets(valueString string, cols []*model.ColumnInfo) ([]string, error) {
+	// FIXME: maybe some values contains '(', ')' or ', '
+	vStr := strings.Trim(strings.Trim(valueString, "("), ")")
```
Comment (csuzhangxc) [inline pkg/dbutil/common.go]: maybe `strings.Trim(valueString, "()")` is enough.
or you can try `TrimPrefix`, `TrimSuffix`.

File: pkg/dbutil/common.go [hunk 6]
```diff
@@ -320,6 +371,138 @@ func GetCRC32Checksum(ctx context.Context, db *sql.DB, schemaName, tableName str
...
+	values := strings.Split(vStr, ", ")
```
Comment (csuzhangxc) [inline pkg/dbutil/common.go:459]: check `len(values)`?

File: pkg/dbutil/common.go [hunk 7]
```diff
@@ -320,6 +371,138 @@ func GetCRC32Checksum(ctx context.Context, db *sql.DB, schemaName, tableName str
...
+func DecodeTimeInBucket(packedStr string, tp byte) (string, error) {
+	var t types.Time
+
+	packed, err := strconv.ParseUint(packedStr, 10, 64)
+	if err != nil {
+		return "", err
+	}
+
+	if packed == 0 {
+		t.Time = types.ZeroTime
```
Comment (csuzhangxc) [inline pkg/dbutil/common.go]: has any effect?

File: pkg/dbutil/common.go [hunk 8]
```diff
@@ -421,3 +604,15 @@ func TableName(schema, table string) string {
...
+// ReplacePlaceholder will use args to replace '?', used for log.
+// tips: make sure the num of "?" is same with len(args)
+func ReplacePlaceholder(str string, args []string) string {
+	/*
+		for example:
+		str is "a > ? AND a < ?", args is {'1', '2'},
+		this function will return "a > '1' AND a < '2'"
+	*/
+	newStr := strings.Replace(str, "?", "'%s'", -1)
+	return fmt.Sprintf(newStr, utils.StringsToInterfaces(args)...)
```
Comment (csuzhangxc) [inline pkg/dbutil/common.go:609]: why need `StringsToInterfaces`?

File: pkg/dbutil/index.go [hunk 9]
```diff
@@ -113,6 +114,49 @@ func FindSuitableIndex(ctx context.Context, db *sql.DB, schemaName string, table
...
+
+	for _, index := range FindAllIndex(tableInfo) {
+		// index will be guaranteed to be visited in order PK -> UK -> IK
+		for _, indexCol := range index.Columns {
+			col := FindColumnByName(tableInfo.Columns, indexCol.Name.O)
+			if _, ok := colsMap[col.Name.O]; ok {
+				continue
+			}
+			colsMap[col.Name.O] = struct{}{}
+			cols = append(cols, col)
```
Comment (csuzhangxc) [inline pkg/dbutil/index.go:153]: if we have two UK `ab` and `ac`, the returned `cols` may `abc` or `acb`, depends on their order in `tableInfo`, right?

File: pkg/dbutil/common.go (hunk 8 above)
Comment (WangXiangUSTC) [reply pkg/dbutil/common.go:609]: if use args will return error `cannot use args (type []string) as type []interface {} in argument to fmt.Sprintf`

File: pkg/dbutil/index.go (hunk 9 above)
Comment (WangXiangUSTC) [reply pkg/dbutil/index.go:153]: yes, it dosen't matter

File: pkg/dbutil/common.go (hunk 4 above)
Comment (WangXiangUSTC) [reply pkg/dbutil/common.go:413]: ok, I will check it

Comment (WangXiangUSTC) [reply pkg/dbutil/common.go:413]: the result of `SHOW STATS_BUCKETS` is transform from mysql. stats_buckets,  mysql. stats_buckets only have tableid, and don't have column name. it is difficult for me to do transform here, I think use `SHOW STATS_BUCKETS`   is more convenient.

**Comment by csuzhangxc** [comment]
LGTM, but CI fail with other problem

**Comment by WangXiangUSTC** [comment]
ci failed because mysql's port is already bind. rebuild and success. Thanks @csuzhangxc 
@GregoryIan  PTAL

File: pkg/dbutil/index.go [hunk 10]
```diff
@@ -113,6 +114,49 @@ func FindSuitableIndex(ctx context.Context, db *sql.DB, schemaName string, table
...
+	copy(indices, tableInfo.Indices)
+	sort.SliceStable(indices, func(i, j int) bool {
+		a := indices[i]
+		b := indices[j]
+		switch {
+		case b.Primary:
+			return false
+		case a.Primary:
+			return true
+		case b.Unique:
```
Comment (IANTHEREAL) [inline pkg/dbutil/index.go:129]: I'm not sure it's right, how to end sort?

File: pkg/dbutil/types.go [hunk 11]
```diff
@@ -23,3 +23,11 @@ func IsFloatType(tp byte) bool {
 
 	return false
 }
+
+// IsTimeType returns true if tp is time type
+func IsTimeType(tp byte) bool {
+	if tp == mysql.TypeDatetime || tp == mysql.TypeTimestamp || tp == mysql.TypeDate {
```
Comment (IANTHEREAL) [inline pkg/dbutil/types.go:29]: what about `TypeNewDate ` and `TypeDuration`?

File: pkg/dbutil/common_test.go [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+func (*testDBSuite) TestTableName(c *C) {
+	testCases := []struct {
+		schema          string
+		table           string
+		expectTableName string
+	}{
+		{
+			"test",
+			"testa",
+			"`test`.`testa`",
```
Comment (IANTHEREAL) [inline pkg/dbutil/common_test.go:53]: add a case like
``` 
`test`.`t``esta` 
```?

File: pkg/dbutil/common.go [hunk 13]
```diff
@@ -320,6 +371,140 @@ func GetCRC32Checksum(ctx context.Context, db *sql.DB, schemaName, tableName str
...
+			values[i] = value
+		}
+	}
+
+	return values, nil
+}
+
+// DecodeTimeInBucket decodes Time from a packed uint64 value.
+// reference: https://github.com/pingcap/tidb/blob/08f0168a6caea0280d6157e5be69f2dc6fd0d5b3/types/time.go#L449
+func DecodeTimeInBucket(packedStr string, tp byte) (string, error) {
```
Comment (IANTHEREAL) [inline pkg/dbutil/common.go]: can we use function of TiDB directly?

**Comment by IANTHEREAL** [comment]
ping @WangXiangUSTC

**Comment by WangXiangUSTC** [comment]
I am doing this now. @GregoryIan

File: pkg/dbutil/types.go (hunk 11 above)
Comment (WangXiangUSTC) [reply pkg/dbutil/types.go:29]: only these three type encoded in tidb bucket. This function's name is not suitable, I rename it.

File: pkg/dbutil/index.go (hunk 10 above)
Comment (WangXiangUSTC) [reply pkg/dbutil/index.go:129]: add test case for this function, it will end sort even if have index with same level.

**Comment by WangXiangUSTC** [comment]
PTAL again @GregoryIan

**Comment by IANTHEREAL** [comment]
LGTM

**Pull request merged**